### PR TITLE
fix overflow bug

### LINKED ISSUE #18357: GracePeriod overflow error
**Bug description**
GracePeriod overflow error

**Expected behavior**
Not showing in logs:
```
2019-10-26T21:49:06.895764Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:06.896749Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:16.896441Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:16.896810Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:26.897183Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:26.897562Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:36.898118Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:36.898658Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:46.898740Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:46.899201Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:56.898917Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:49:56.899447Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:06.899699Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:06.900406Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:16.900253Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:16.900729Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:26.901361Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:26.901989Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:36.901955Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:36.902465Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:46.902435Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:46.902907Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:56.903112Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:50:56.903560Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:06.903608Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:06.904102Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:16.904421Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:16.904814Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:26.904874Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:26.905402Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:36.906169Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:36.906681Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.                                                                                 
2019-10-26T21:51:46.906579Z    warn    gracePeriod (8760h0m0s * 0.500000) = -744h5m44.073709551s is less than minGracePeriod 10m0s. Apply minGracePeriod.
```

**Steps to reproduce the bug**
Have a high "certLifeTime" value. See https://github.com/istio/istio/blob/d5964882ca1d0932623a233aa135aaba66b628eb/security/pkg/k8s/chiron/controller.go#L305.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
client version: 1.3.3
nodeagent version:
citadel version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a
galley version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a
policy version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a
telemetry version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a
sidecar-injector version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a
ingressgateway version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a
pilot version: 1.5-alpha.ea451fcda84854b21bdde53d5b3f027dad01d58a

Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2-k3s.1", GitCommit:"b8b17ba55f20e590df507fce333dfee13ab438c6", GitTreeState:"clean", BuildDate:"2019-10-16T05:17Z", GoVersion:"go1.13.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2-k3s.1", GitCommit:"b8b17ba55f20e590df507fce333dfee13ab438c6", GitTreeState:"clean", BuildDate:"2019-10-16T05:17Z", GoVersion:"go1.13.3", Compiler:"gc", Platform:"linux/amd64"}

### PR DESCRIPTION
**Pull request by inteon**
Fixes #18357

### CODE REVIEW

**Comment by inteon** [comment]
@googlebot I signed it!

File: security/pkg/k8s/chiron/controller.go [hunk 1]
```diff
@@ -300,7 +300,7 @@ func (wc *WebhookController) scrtUpdated(oldObj, newObj interface{}) {
 	certLifeTimeLeft := time.Until(cert.NotAfter)
 	certLifeTime := cert.NotAfter.Sub(cert.NotBefore)
 	// Because time.Duration only takes int type, multiply gracePeriodRatio by 1000 and then divide it.
-	gracePeriod := time.Duration(wc.gracePeriodRatio*1000) * certLifeTime / 1000
+	gracePeriod := certLifeTime / 1000 * time.Duration(wc.gracePeriodRatio*1000)
```
Comment (incfly) [inline security/pkg/k8s/chiron/controller.go]: whats the typical value range for gracePeriodratio and certLifeTime, can we just do certLifeTime * gracePeriodratio?

Comment (lei-tang) [reply security/pkg/k8s/chiron/controller.go]: The fix in this PR (i.e., gracePeriod := certLifeTime / 1000 * time.Duration(wc.gracePeriodRatio*1000)) still has an underflow problem, e.g., when certLifeTime := time.Duration(10*time.Nanosecond), the underflow problem occurs with the output as follows.
~$ go run main.go
gracePeriod (10ns * 0.500000) = 0s is less than minGracePeriod 2ns. Apply minGracePeriod.

The test program to show the underflow is as follows:
package main

import (
	"fmt"
	"time"
)

func main() {
	gracePeriodRatio := 0.5
	certLifeTime := time.Duration(10*time.Nanosecond)
	minGracePeriod := 2 * time.Nanosecond

	// The following case will underflow, e.g., when certLifeTime certLifeTime := time.Duration(10*time.Nanosecond) 
	gracePeriod := certLifeTime / 1000 * time.Duration(gracePeriodRatio*1000)
	
	if gracePeriod < minGracePeriod {
		fmt.Printf("gracePeriod (%v * %f) = %v is less than minGracePeriod %v. Apply minGracePeriod.\n",
			certLifeTime, gracePeriodRatio, gracePeriod, minGracePeriod)
		gracePeriod = minGracePeriod
	} else {
		fmt.Printf("gracePeriod (%v * %f) = %v\n",
			certLifeTime, gracePeriodRatio, gracePeriod)
	}
}

Comment (lei-tang) [reply security/pkg/k8s/chiron/controller.go]: The PR https://github.com/istio/istio/pull/18360 fixes an overflow problem at a different function (i.e., GetWaitTime()) and refactors the code here to reuse the GetWaitTime() function.

Comment (lei-tang) [reply security/pkg/k8s/chiron/controller.go]: The PR https://github.com/istio/istio/pull/18360 has been merged. This PR can be closed.

**Pull request closed without merging**